src/main/java/com/avispl/symphony/dal/communicator/barco/e2gen2/BarcoE2Communicator.java

```diff
@@ -690,18 +725,51 @@ private void updateSourcePropertiesForScreenDest(JsonNode destContent, SourcePro
 		JsonNode layers = destContent.get(BarcoE2Constant.LAYERS);
 		List<Integer> listSourceIndexes = new ArrayList<>();
 		for (int j = 0; j < layers.size(); j++) {
+			if (layers.get(j).get(BarcoE2Constant.LAST_SRC_IDX) == null) {
+				continue;
+			}
 			int sourceIndex = layers.get(j).get(BarcoE2Constant.LAST_SRC_IDX).asInt();
-			if (sourceIndex != -1) {
+			int pgmMode = layers.get(j).get(BarcoE2Constant.PGM_MODE).asInt();
+			int linkDestId = layers.get(j).get(BarcoE2Constant.LINK_DEST_ID).asInt();
+			if (sourceIndex != -1 && pgmMode == 1 && sourceIdToNameMap.get(sourceIndex) != null && linkDestId == -1) {
 				listSourceIndexes.add(sourceIndex);
 			}
 		}
-		Set<Integer> uniqueList = new HashSet<>(listSourceIndexes);
-		sourceProperties.numberOfSource = uniqueList.size();
+		sourceProperties.numberOfSource = listSourceIndexes.size();
 		if (!listSourceIndexes.isEmpty()) {
 			sourceProperties.currentSourceName = sourceIdToNameMap.get(listSourceIndexes.get(0));
+		} else {
+			sourceProperties.currentSourceName = BarcoE2Constant.NONE;
 		}
 	}
 
+	/**
+	 * Update source properties for super screen destination
+	 *
+	 * @param destContent content of super screen destination
+	 * @param sourceProperties SourceProperties class
+	 */
+	private void updateSourcePropertiesForSuperScreenDest(JsonNode destContent, SourceProperties sourceProperties) {
+		JsonNode layers = destContent.get(BarcoE2Constant.LAYERS);
+		List<Integer> listSourceIndexes = new ArrayList<>();
+		for (int j = 0; j < layers.size(); j++) {
+			if (layers.get(j).get(BarcoE2Constant.LAST_SRC_IDX) == null) {
```

Comment:
layers.get(j) can be extracted and reused, 
also please check for potential NPEs

Reply:
Fixed